### PR TITLE
ENH: Refactor LoRA bnb layers for faster initialization

### PR DESCRIPTION
**Pull request by BenjaminBossan**
Partly addresses #896

## Description

After speeding up normal LoRA layer initialization, this PR improves initialization speed of bnb LoRA layers.

The method to achieve this is different from the one used before, namely this time the base layer is stored as a reference on the LoRA layer. This allows us to avoid calling `__init__` on the bnb layer, which is what is slow.

## Notes

We cannot use the same method as for the normal LoRA layers, (i.e. calling the super class's `__init__` with meta device) because the bnb layers have extra logic that still creates unnecessary weights (this is something that could potentially be fixed by bnb).

However, the way used here could also be a solution to the normal layers, so if we want to have consistency, the normal layers could be refactored to use the same approach.

Another small advantage of this approach is that we can now correctly handle any `*args, **kwargs` that are passed to `forward`. If we had this consistently in all layers, it would allow us to handle extra arguments like `lora_scale` in the future.

Interestingly, even though we now save the base layer as a reference, which results in a different `state_dict`, the existing models can still be loaded successfully. This is because the adapter `state_dict` is not affected by the change, so users can still load their existing adapters.

The only problem would occur if users dump the whole model, i.e. base model and adapter, using `torch.save` and then trying to load with `torch.load`. For those users, we could theoretically provide a script to convert the `state_dict` (i.e. renaming some keys).

To ensure that the old adapters can still be loaded successfully, I'm working at the same time on adding regression tests. I'll create a separate PR for those to avoid blowing up this one.

## Tests

I ran a test on `bloomz-1b1` for how long it takes to create the `PeftModel`, the results are:

8bit: 1108.34 ms > 26.82 ms
4bit: 1101.96 ms > 23.69 ms

Script was:

```python
import time
from contextlib import contextmanager

import torch
from transformers import AutoModelForCausalLM, BitsAndBytesConfig
from peft import LoraConfig, get_peft_model

@contextmanager
def timed(msg):
    start = time.perf_counter()
    yield
    end = time.perf_counter()
    print(f"{msg} took {1000 * (end - start):.2f} ms")

torch.manual_seed(1000)
model_id = "bigscience/bloomz-1b1"
test_8bit = True

if test_8bit:
    model = AutoModelForCausalLM.from_pretrained(
        model_id,
        load_in_8bit=True,
    )
    config = LoraConfig(
        r=8,
        init_lora_weights=False,
    )
else:
    bnb_config = BitsAndBytesConfig(
        load_in_4bit=True,
        bnb_4bit_use_double_quant=False,
        bnb_4bit_compute_type=torch.float32,
    )
    model = AutoModelForCausalLM.from_pretrained(
        model_id,
        quantization_config=bnb_config,
        torch_dtype=torch.float32,
    )
    config = LoraConfig(
        r=8,
        init_lora_weights=False,
    )

with timed("creating peft model"):
    get_peft_model(model, config)
```

### CODE REVIEW

File: src/peft/tuners/lora/bnb.py [hunk 1]
```diff
@@ -62,39 +51,55 @@ def merge(self):
...
-                        self.weight.data, to_order=self.state.formatB
-                    )
-                out32, Sout32 = bnb.functional.igemmlt(im, self.state.CxB, Sim, self.state.SB)
-                output = bnb.functional.mm_dequant(out32, Sout32, SCim, self.state.SCB, bias=None).t()
                 w_data = output.to(lora_data.dtype).to(lora_data.device) + lora_data
-                self.weight = bnb.nn.Int8Params(
-                    w_data.to("cpu"), requires_grad=False, has_fp16_weights=self.weight.has_fp16_weights
-                ).to(self.weight.device)
-                self.state.reset_grads()
+                self.base_layer.weight = bnb.nn.Int8Params(
```
Comment (pacman100) [inline src/peft/tuners/lora/bnb.py:78]: I believe that on line 77, the output needs to be the newer base weights but they are the initial ones itself leading to this being incorrect when multiple adapters are active.

File: src/peft/tuners/lora/model.py [hunk 2]
```diff
@@ -199,12 +199,23 @@ def _replace_module(parent, child_name, new_module, child):
...
+        # TODO: layers with base_layer don't need the weight to be copied, as they have a reference already
+        if not hasattr(new_module, "base_layer"):
+            new_module.weight = child.weight
+            if hasattr(child, "bias"):
+                new_module.bias = child.bias
 
         if getattr(child, "state", None) is not None:
-            new_module.state = child.state
+            if not hasattr(new_module, "base_layer"):
+                # TODO is setting state even necessary in any situation?
```
Comment (pacman100) [inline src/peft/tuners/lora/model.py:217]: Yes, else the weights won't be initialized properly and bnb layers would give wrong results.

**Review by pacman100 — COMMENTED** [review]
Thank you @BenjaminBossan for speeding up the creation of quantized peft models 🚀. I like the approach and it simplifies and makes it more extendable. Left a couple of comments.

File: src/peft/tuners/lora/bnb.py (hunk 1 above)
Comment (BenjaminBossan) [reply src/peft/tuners/lora/bnb.py:78]: To clarify, do you mean that the `output`, as computed in line 68, should instead be computed inside the loop, i.e. it changes on each iteration? Would that also mean that all the other lines above that, like computing `out32` and `Sout32`, need to be computed inside the loop?

To me, it looked like these values wouldn't change on each loop, but maybe I'm missing some subtle dependency.

File: src/peft/tuners/lora/model.py (hunk 2 above)
Comment (BenjaminBossan) [reply src/peft/tuners/lora/model.py:217]: Okay, thanks, so it means the code should rather be:

```python
if hasattr(new_module, "base_layer"):
    new_module.base_layer.state = child.state
else:
    new_module.state = child.state
```

Do I understand that correctly?

Comment (BenjaminBossan) [reply src/peft/tuners/lora/model.py]: Done

File: src/peft/tuners/lora/bnb.py (hunk 1 above)
Comment (BenjaminBossan) [reply src/peft/tuners/lora/bnb.py:79]: Done

**Review by pacman100 — APPROVED** [review]
Thank you @BenjaminBossan, good to merge after resolving the conflicts.
